lib/compress/zstd_compress_superblock.c

```diff
@@ -427,17 +433,43 @@ static size_t countLiterals(seqStore_t const* seqStore, const seqDef* sp, size_t
     for (n=0; n<seqCount; n++) {
         total += ZSTD_getSequenceLength(seqStore, sp+n).litLength;
     }
-    DEBUGLOG(5, "countLiterals for %zu sequences from %p => %zu bytes", seqCount, (const void*)sp, total);
+    DEBUGLOG(6, "countLiterals for %zu sequences from %p => %zu bytes", seqCount, (const void*)sp, total);
     return total;
 }
 
+#define BYTESCALE 256
+
+static size_t sizeBlockSequences(const seqDef* sp, size_t nbSeqs,
+                size_t targetBudget, size_t avgLitCost, size_t avgSeqCost,
+                int firstSubBlock)
```

Comment:
It would be pretty easy to modify this function to count the total uncompressed bytes in the sequences, and not cut a block as long as the uncompressed bytes is less than `targetCBlockSize`. That would be a simple way to guarantee that we don't violate compress bound.

I think it would also help keep block sizes even, since if the uncompressed size is less than `targetCBlockSize`, then the compressed size will be as well.

Given PR #1947, it likely isn't 100% necessary, but I think it is a nice sanity check, and should help improve compression ratio in edge cases.

Reply:
> That would be a simple way to guarantee that we don't violate compress bound.

If `compressBound()` violation is the concern here,
let's underline that the total nb of blocks produced is limited, bounded by `nbSubBlocks` : 
this code never generates more than `nbSubBlocks` blocks.

Since `nbSubBlocks = estimateCSize / targetCBlockSize`,
and `targetCBlockSize` is bounded by a minimum of `CBLOCK_TARGET_SIZE_MIN` (`1300` bytes currently),
`nbSubBlocks` cannot define a nb of blocks > 1 block per 1 KB of input data
hence it cannot defeat `compressBound()`,

unless the source block itself is < 1 KB as a consequence of user-controlled `flush()`,
in which case the `compressBound()` guarantee is out of scope, and there will be no split.

Reply:
Another way to look at it is that `compressBound()` is only a concern when blocks are not compressed.
In this case, note that this is controlled just below,
and if a sub-block is not compressed, it's not generated immediately => it's actually coalesced with the next one (effectively creating a larger sub-block). This process continues until the coalesced sub-block becomes compressible, or until reaching the end of the super-block. In which case, a single (final) uncompressed sub-block is produced per superblock.
So it can never degenerates into producing many small uncompressed sub-blocks.

Reply:
I still think that it is worth controlling the case where `uncompressedSize < targetCBlockSize`.

I know that compress bound really shouldn't be violated. First, by the fact that `nbSubBlocks` is supposed to not be too big. Second by the final check at the end that will emit an uncompressed block if `dstSize_tooSmall` or we don't shrink by enough.

However, I think this is a simple check that we can add that will stop us from emitting too-small blocks in some edge cases, and will help block consistency.

Reply:
>  if uncompressedSize < targetCBlockSize, we don't want to emit a block anyway.

But isn't that already taken care of by the following test : 
`if (cSize > 0 && cSize < decompressedSize) {`